javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt

```diff
@@ -0,0 +1,178 @@
+import arcs.core.entity.DummyEntity
+import arcs.core.entity.HandleSpec
+import arcs.core.entity.InlineDummyEntity
+import arcs.core.entity.ReadWriteCollectionHandle
+import arcs.core.entity.ReadWriteSingletonHandle
+import arcs.core.entity.awaitReady
+import arcs.core.host.EntityHandleManager
+import arcs.core.storage.DriverFactory
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StoreWriteBack
+import arcs.core.storage.driver.RamDisk
+import arcs.core.storage.keys.DatabaseStorageKey
+import arcs.core.storage.keys.RamDiskStorageKey
+import arcs.core.storage.keys.VolatileStorageKey
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+import arcs.core.storage.testutil.WriteBackForTesting
+import arcs.core.testutil.handles.dispatchStore
+import arcs.core.util.testutil.LogRule
+import arcs.jvm.host.JvmSchedulerProvider
+import arcs.jvm.util.testutil.FakeTime
+import arcs.sdk.android.storage.AndroidDriverAndKeyConfigurator
+import com.google.common.truth.Truth.assertThat
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.coroutineContext
+import kotlinx.coroutines.runBlocking
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/** Tests for [StorageServiceManager]. */
+@Suppress("UNCHECKED_CAST", "EXPERIMENTAL_API_USAGE")
+@RunWith(AndroidJUnit4::class)
+class StorageServiceEndToEndTest {
+    @get:Rule
+    val log = LogRule()
+
+    private suspend fun buildManager() =
+        StorageServiceManager(coroutineContext, ConcurrentHashMap())
+    private val time = FakeTime()
+    private val scheduler = JvmSchedulerProvider(EmptyCoroutineContext).invoke("test")
+    private val ramdiskKey = ReferenceModeStorageKey(
+        backingKey = RamDiskStorageKey("backing"),
+        storageKey = RamDiskStorageKey("container")
+    )
+    private val arcId = ArcId.newForTest("foo")
+    private val volatileKey = ReferenceModeStorageKey(
+        backingKey = VolatileStorageKey(arcId, "backing"),
+        storageKey = VolatileStorageKey(arcId, "container")
+    )
+    private val databaseKey = ReferenceModeStorageKey(
+        backingKey = DatabaseStorageKey.Persistent("backing", DummyEntity.SCHEMA_HASH),
+        storageKey = DatabaseStorageKey.Persistent("container", DummyEntity.SCHEMA_HASH)
+    )
+
+    @Before
+    fun setUp() {
+        StoreWriteBack.writeBackFactoryOverride = WriteBackForTesting
+        AndroidDriverAndKeyConfigurator.configure(ApplicationProvider.getApplicationContext())
+        SchemaRegistry.register(DummyEntity.SCHEMA)
+        SchemaRegistry.register(InlineDummyEntity.SCHEMA)
+    }
+
+    @After
+    fun tearDown() {
+        WriteBackForTesting.clear()
+        scheduler.cancel()
+        RamDisk.clear()
+        DriverFactory.clearRegistrations()
+    }
+
+    @Test
+    fun writeThenReadOfInlineSingletonInCollectionOnDatabase() = runBlocking {
```

Comment:
Nit: place `_`'s strategically to make the test methods easier to read.

For example, this could probably be something like:

```
fun writeThenRead_inlineSingleton_inCollection_onDatabase() = ...
```